### PR TITLE
SJQ2T-21: Forgot password functionality

### PR DESCRIPTION
**Pull request by DraganMitrasinovicHtec**
Added the forgot password functionality, user can now request a password reset email.
After getting the email, he will be able to go to that page and enter a new password twice to reset it.
Still need to integrate it with the frontend guys once they create the page.
There is some reformatting and variable setting left but we will clean up the whole project soon.

### CODE REVIEW

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -64,14 +64,41 @@ public void sendVerificationEmail(String emailAddress) {
...
+
+    String confirmationLink = jwtUtils.getPasswordResetLink() + token;
+    String emailBody = jwtUtils.buildPasswordResetEmail(email, confirmationLink);
+    Email emailToSend = new Email(email, SUBJECT_FOR_PASSWORD_RESET, emailBody, true);
+    emailService.send(emailToSend);
+  }
+
+  @Override
+  public String resetPassword(String token, String password) {
+    boolean validateToken = jwtUtils.validateJwtToken(token);
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java]: Maybe method validateJwtToken can throw exception if token is not valid and to be void, instead boolean ?

File: backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java [hunk 2]
```diff
@@ -132,4 +138,77 @@ public String buildEmail(String email, String link) {
         + "\n"
         + "</div></div>";
   }
+
+  public String buildPasswordResetEmail(String email, String link) {
+    return "<div style=\"font-family:Helvetica,Arial,sans-serif;font-size:16px;margin:0;color:#0b0c0c\">\n"
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:143]: Is there a way to move this huge String in some file, and to read it from file ?

Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:143]: There should be, we plan to do the same with the verification email.

**Review by markobarackovhtec — APPROVED** [review]

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java (hunk 1 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java]: That makes sense, will make the method look cleaner as it won't have the if statement and exception throwing. Will change

File: backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java (hunk 2 above)
Comment (markobarackovhtec) [reply backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:143]: Then create a ticket for that, so that you dont forget it :)

Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:143]: I made sure to add it as part of the refactor task checklist

**Comment by DraganMitrasinovicHtec** [comment]
Changed validateToken to void
Deleted now irrelevant tests
Changed url path naming to hyphens
Writing a comment because github doesn't show the full title

File: backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  public void validateJwtToken(String token) {
+    try {
+      Verification verification = JWT.require(Algorithm.HMAC256(jwtSecret));
+      Date expiresAt = verification.build().verify(token).getExpiresAt();
+      JWTVerifier verifier = verification.acceptExpiresAt(expiresAt.getTime()).build();
+      verifier.verify(token);
+    } catch (JWTVerificationException e) {
+      log.error("Invalid JWT token: {}", e.getMessage());
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:53]: Throw exception here, i believe that you need to return error to end user if token is not valid ?

File: backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    } catch (IllegalArgumentException e) {
+      log.error("JWT claims string is empty: {}", e.getMessage());
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:56]: same

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -86,14 +111,12 @@ public UserDto findUserByEmail(String email) {
 
   @Override
   public String confirmEmail(String token) {
-    boolean validateToken = jwtEmailVerificationUtils.validateJwtToken(token);
-    if (validateToken) {
-      String email = jwtEmailVerificationUtils.getEmailFromJwtEmailVerificationToken(token);
+    jwtUtils.validateJwtToken(token);
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java:113]: Not formatted correctly

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -64,14 +64,39 @@ public void sendVerificationEmail(String emailAddress) {
...
+
+    String confirmationLink = jwtUtils.getPasswordResetLink() + token;
+    String emailBody = jwtUtils.buildPasswordResetEmail(email, confirmationLink);
+    Email emailToSend = new Email(email, SUBJECT_FOR_PASSWORD_RESET, emailBody, true);
+    emailService.send(emailToSend);
+  }
+
+  @Override
+  public String resetPassword(String token, String password) {
+    jwtUtils.validateJwtToken(token);
```
Comment (markobarackovhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java:91]: Not formatted correctly

File: backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java (hunk 3 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/util/JwtUtils.java:53]: Makes sence, so the front-end can show them a pop-up or something.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java (hunk 5 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java:113]: My bad, thought I turned on autoformat on commit.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/UserServiceImpl.java:91]: Will clean it up

**Review by markobarackovhtec — APPROVED** [review]

**Pull request merged**